### PR TITLE
synchronize: quotes around arguments

### LINKED ISSUE #24: Synchronize module not recognizing remote ssh key
##### SUMMARY

I am attempting to synchronize in push mode a couple of directories from one remote host (A) to a second remote host (B). The rsync command shown in ansible's output works as expected when run on remote host A.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Synchronize module

##### ANSIBLE VERSION

```
ansible 2.9.9
  config file = /home/evan/repos/infrastructure_ansible/ansible.cfg
  configured module search path = [u'/home/evan/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.17 (default, Apr 15 2020, 17:20:14) [GCC 7.5.0]

```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

Ansible host: Ubuntu 18.04
Remote hosts: Ubuntu 18.04

##### STEPS TO REPRODUCE

```
- synchronize:  # push from alpha to others
    mode: push
    src: "/home/{{ ansible_user }}/robot_code/cloud_sync_siam"
    dest: "/home/{{ ansible_user }}/robot_code"
    recursive: 'yes'
    delete: 'yes'
    rsync_opts: --rsh="/usr/bin/ssh -o PasswordAuthentication=no -i /home/{{ ansible_user }}/.ssh/team_van_rsa"
  become: "{{ ansible_user }}"
  delegate_to: "{{ inventory_host }}"
```

##### EXPECTED RESULTS

A successful rsync from the delegated remote host to remote host B

##### ACTUAL RESULTS

It appears that the ssh key to be used for connecting from remote host A to remote host B cannot be found. However, the given rsync command runs as expected when run as {{ ansible_user }} on the delegated host (after including quotes around the --out-format).

```
TASK [abii-updater : synchronize] *******************************************************************************************
fatal: [thruway-dev-bravo -> 13.68.168.84]: FAILED! => changed=false 
  cmd: /usr/bin/rsync --delay-updates -F --compress --delete-after --archive --rsh="/usr/bin/ssh -o PasswordAuthentication=no -i /home/azureuser/.ssh/team_van_rsa" --out-format=<<CHANGED>>%i %n%L /home/azureuser/robot_code/cloud_sync_siam azureuser@20.185.12.157:/home/azureuser/robot_code
  invocation:
    module_args:
      _local_rsync_password: null
      _local_rsync_path: rsync
      _substitute_controller: false
      archive: true
      checksum: false
      compress: true
      copy_links: false
      delete: true
      dest: azureuser@20.185.12.157:/home/azureuser/robot_code
      dest_port: 22
      dirs: false
      existing_only: false
      group: null
      link_dest: null
      links: null
      mode: push
      owner: null
      partial: false
      perms: null
      private_key: /home/evan/.ssh/team_van_rsa
      recursive: true
      rsync_opts:
      - --rsh="/usr/bin/ssh -o PasswordAuthentication=no -i /home/azureuser/.ssh/team_van_rsa"
      rsync_path: null
      rsync_timeout: 0
      set_remote_user: true
      src: /home/azureuser/robot_code/cloud_sync_siam
      ssh_args: null
      times: null
      verify_host: false
  msg: |-
    rsync: Failed to exec /usr/bin/ssh -o PasswordAuthentication=no -i /home/azureuser/.ssh/team_van_rsa: No such file or directory (2)
    rsync error: error in IPC code (code 14) at pipe.c(85) [sender=3.1.2]
    rsync: connection unexpectedly closed (0 bytes received so far) [sender]
    rsync error: error in IPC code (code 14) at io.c(235) [sender=3.1.2]
  rc: 14
```

### PR DESCRIPTION
**Pull request by flybyray**
fix quoting for specific cmd arguments

Fixes:
- #24
- #190 

ISSUE TYPE:
- Bugfix Pull Request

COMPONENT NAME:
  module: synchronize

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @quidame @saito-hideki @aminvakil Could you please review this? Thanks in advance.

File: plugins/modules/synchronize.py [hunk 1]
```diff
@@ -548,10 +548,10 @@ def main():
             ssh_cmd_str = ' '.join(shlex_quote(arg) for arg in ssh_cmd)
             if ssh_args:
                 ssh_cmd_str += ' %s' % ssh_args
-            cmd.append('--rsh=%s' % ssh_cmd_str)
+            cmd.append(shlex_quote('--rsh=%s' % ssh_cmd_str))
 
     if rsync_path:
-        cmd.append('--rsync-path=%s' % rsync_path)
```
Comment (Akasurde) [inline plugins/modules/synchronize.py:554]: Would it be possible to add tests for this change?

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by quidame** [comment]
hi @flybyray , thanks for fixing this !
Currently integration tests don't fail because they are disabled in `tests/integration/targets/synchronize/aliases`. Can you enable them and also add tests related to this PR ?

File: plugins/modules/synchronize.py (hunk 1 above)
Comment (flybyray) [reply plugins/modules/synchronize.py:554]: There is already a test for it.
https://github.com/ansible-collections/ansible.posix/blob/main/tests/integration/targets/synchronize/tasks/main.yml#L154-L166

As @quidame mentioned later the integration tests were disabled sometime ago. I now enabled them. localy with docker it worked

```
TASK [synchronize : synchronize files using rsync_path (issue#7182)] ***********
changed: [testhost] => {"changed": true, "cmd": "/usr/bin/rsync --delay-updates -F --compress --archive '--rsync-path=sudo rsync' '--out-format=<<CHANGED>>%i %n%L' /root/ansible_collections/ansible/posix/tests/output/.tmp/output_dir/foo.txt /root/ansible_collections/ansible/posix/tests/output/.tmp/output_dir/foo.rsync_path", "msg": ">f+++++++++ foo.txt\n", "rc": 0, "stdout_lines": [">f+++++++++ foo.txt"]}

TASK [synchronize : assert] ****************************************************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [synchronize : Cleanup] ***************************************************
changed: [testhost] => (item=foo.rsync_path) => {"ansible_loop_var": "item", "changed": true, "item": "foo.rsync_path", "path": "/root/ansible_collections/ansible/posix/tests/output/.tmp/output_dir/foo.rsync_path", "state": "absent"}
```

Comment (saito-hideki) [reply plugins/modules/synchronize.py:554]: Hi @Akasurde @flybyray,
Based on the code changes, I think we  probably need to append the integration test like below:
```YAML
- name: command-line options are quoted correctly (issue#24 and issue#190)
  synchronize:
    src: "{{output_dir}}/foo.txt"
    dest: "dummy:{{output_dir}}/foo.rsync_path_1"
  register: sync_result
  ignore_errors: true
- assert:
    that:
    - sync_result.cmd is search("'--rsh=.*'")
    - sync_result.cmd is search("'--out-format=.*'")
```

**Review by saito-hideki — CHANGES_REQUESTED** [review]
Based on the code changes, I think we probably need to append the integration test like below:
```YAML
- name: command-line options are quoted correctly (issue#24 and issue#190)
  synchronize:
    src: "{{output_dir}}/foo.txt"
    dest: "dummy:{{output_dir}}/foo.rsync_path_1"
  register: sync_result
  ignore_errors: true
- assert:
    that:
    - sync_result.cmd is search("'--rsh=.*'")
    - sync_result.cmd is search("'--out-format=.*'")
```

**Comment by flybyray** [comment]
> Based on the code changes, I think we probably need to append the integration test like below:
> 
> ```yaml
> - name: command-line options are quoted correctly (issue#24 and issue#190)
>   synchronize:
>     src: "{{output_dir}}/foo.txt"
>     dest: "dummy:{{output_dir}}/foo.rsync_path_1"
>   register: sync_result
>   ignore_errors: true
> - assert:
>     that:
>     - sync_result.cmd is search("'--rsh=.*'")
>     - sync_result.cmd is search("'--out-format=.*'")
> ```

thanks. it should reference somehow the issues with their reported problems?

Can someone please have a look at those two changes:

https://github.com/ndgit/ansible.posix/blame/fix-24-and-190/plugins/modules/synchronize.py#L604
>> https://github.com/ndgit/ansible.posix/commit/0118bf0cb9256637684db3f18cbd012d3851065e#diff-3332cfa3b3869c49a3c44104c6662f6e08d303f01ca9491981a2ba38e333a560L604

https://github.com/ndgit/ansible.posix/blame/fix-24-and-190/plugins/modules/synchronize.py#L607
>> https://github.com/ndgit/ansible.posix/commit/0118bf0cb9256637684db3f18cbd012d3851065e#diff-3332cfa3b3869c49a3c44104c6662f6e08d303f01ca9491981a2ba38e333a560L607

I think they would not be required if the `run_command` from basic module would work as expected. It looks to me that there is most probably another issue, which should not be solved here.

with my change the command will be run differently: 
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1856-L1858

The second would be the case with my change an not the first:
> If args is a string and use_unsafe_shell=False it will split args to a list and run with shell=False

The implementation might beinefficient. Fist building a string from a list, than building a list from string.

**Review by saito-hideki — APPROVED** [review]
@flybyray Thank you for your comment. As you pointed out, my integration test example is not essential for this PR as it only tests the behavior of shlex_quote().
I also confirmed the integration test works fine now without '--allow-disabled'.

**Review by gravesm — APPROVED** [review]

**Review by Akasurde — APPROVED** [review]

**Comment by alisonatwork** [comment]
@gravesm @Akasurde @saito-hideki in case you didn't notice the bug report - this change has introduced a bug that breaks (formerly) correct and working synchronize tasks. Please can you take a look at the bug report #449 and suggest what to do from here.

**Pull request merged**